File: Sources/ErkmoAnalytics/ErkmoAnalytics.swift
import Foundation
import UIKit

public final class ErkmoAnalytics {
    public static let shared = ErkmoAnalytics()

    private var siteId: String = ""
    private var apiUrl: String = "https://t.erkmo.com/track"
    private var sessionId: String = ""
    private let sdkVersion = "ios-0.1"

    private init() {}

    public func configure(siteId: String, apiUrl: String = "https://t.erkmo.com/track") {
        self.siteId = siteId
        self.apiUrl = apiUrl
        self.sessionId = ErkmoAnalytics.loadOrCreateSessionId()
    }

    public func track(event: String, properties: [String: Any] = [:]) {
        send(event: event, properties: properties, screenName: nil)
    }

    public func screen(_ name: String, properties: [String: Any] = [:]) {
        send(event: "screen_view", properties: properties, screenName: name)
    }

    private func send(event: String, properties: [String: Any], screenName: String?) {
        guard !siteId.isEmpty, let url = URL(string: apiUrl) else { return }

        var payload: [String: Any] = [
            "site_id": siteId,
            "event": event,
            "timestamp": Int(Date().timeIntervalSince1970 * 1000),
            "session_id": sessionId,
            "platform": "ios",
            "sdk_version": sdkVersion,
            "app_name": Bundle.main.object(forInfoDictionaryKey: "CFBundleName") as? String ?? "iOS App",
            "app_id": Bundle.main.bundleIdentifier ?? "unknown",
            "app_version": Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "0.0.0",
            "app_build": Bundle.main.object(forInfoDictionaryKey: "CFBundleVersion") as? String ?? "0",
            "device_id": UIDevice.current.identifierForVendor?.uuidString ?? "",
            "screen_name": screenName ?? ""
        ]

        payload["properties"] = properties

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: payload, options: [])

        URLSession.shared.dataTask(with: request).resume()
    }

    private static func loadOrCreateSessionId() -> String {
        let key = "erkmo_session_id"
        let defaults = UserDefaults.standard
        if let existing = defaults.string(forKey: key), !existing.isEmpty {
            return existing
        }
        let newId = "session_\(UUID().uuidString)"
        defaults.set(newId, forKey: key)
        return newId
    }
}

File: sdk/android/src/main/java/com/erkmo/analytics/ErkmoAnalytics.kt
package com.erkmo.analytics

import android.content.Context
import android.provider.Settings
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.util.UUID

object ErkmoAnalytics {
    private var siteId: String = ""
    private var apiUrl: String = "https://t.erkmo.com/track"
    private var sessionId: String = ""
    private var kidsMode: Boolean = false
    private const val sdkVersion = "android-0.1.1"

    fun init(context: Context, siteId: String, apiUrl: String = "https://t.erkmo.com/track", kidsMode: Boolean = false) {
        this.siteId = siteId
        this.apiUrl = apiUrl
        this.kidsMode = kidsMode
        this.sessionId = loadOrCreateSessionId(context)
    }

    fun track(context: Context, event: String, properties: Map<String, Any> = emptyMap()) {
        send(context, event, properties, null)
    }

    fun screen(context: Context, name: String, properties: Map<String, Any> = emptyMap()) {
        send(context, "screen_view", properties, name)
    }

    private fun send(context: Context, event: String, properties: Map<String, Any>, screenName: String?) {
        if (siteId.isBlank()) return
        val payload = JSONObject()
        payload.put("site_id", siteId)
        payload.put("event", event)
        payload.put("timestamp", System.currentTimeMillis())
        payload.put("session_id", sessionId)
        payload.put("platform", "android")
        payload.put("sdk_version", sdkVersion)
        payload.put("app_name", context.applicationInfo.loadLabel(context.packageManager).toString())
        payload.put("app_id", context.packageName)
        payload.put("app_version", context.packageManager.getPackageInfo(context.packageName, 0).versionName ?: "0.0.0")
        payload.put("app_build", context.packageManager.getPackageInfo(context.packageName, 0).longVersionCode.toString())
        val deviceId = if (kidsMode) "" else (Settings.Secure.getString(context.contentResolver, Settings.Secure.ANDROID_ID) ?: UUID.randomUUID().toString())
        payload.put("device_id", deviceId)
        if (!screenName.isNullOrBlank()) payload.put("screen_name", screenName)

        val props = JSONObject()
        for ((k, v) in properties) {
            props.put(k, v)
        }
        if (kidsMode) {
            props.put("kids_mode", true)
        }
        payload.put("properties", props)

        Thread {
            try {
                val connection = (URL(apiUrl).openConnection() as HttpURLConnection).apply {
                    requestMethod = "POST"
                    setRequestProperty("Content-Type", "application/json")
                    doOutput = true
                }
                OutputStreamWriter(connection.outputStream).use { it.write(payload.toString()) }
                connection.inputStream.close()
                connection.disconnect()
            } catch (_: Exception) {
                // Ignore network errors to avoid crashing the host app
            }
        }.start()
    }

    private fun loadOrCreateSessionId(context: Context): String {
        val prefs = context.getSharedPreferences("erkmo_analytics", Context.MODE_PRIVATE)
        val existing = prefs.getString("session_id", null)
        if (!existing.isNullOrBlank()) return existing
        val newId = "session_${UUID.randomUUID()}"
        prefs.edit().putString("session_id", newId).apply()
        return newId
    }
}
